lib/Transforms/Utils/PromoteMemoryToRegister.cpp

```diff
@@ -562,6 +574,26 @@ void PromoteMem2Reg::run() {
     // analogous to finding the 'uses' and 'definitions' of each variable.
     Info.AnalyzeAlloca(AI);
 
+    // For allocas that have at least one store, make sure the block that holds
+    // the lifetime.start call is treated as a 'definition'.
+    // Ignore allocas that are only used in one block, even if the lifetime
+    // start marker is in a different one.
+    if (LifetimeStartBB && !Info.DefiningBlocks.empty() && !Info.OnlyUsedInOneBlock) {
+      bool AnyStoreInStartBB = false;
+      for (BasicBlock *BB : Info.DefiningBlocks) {
+        if (BB == LifetimeStartBB) {
+          AnyStoreInStartBB = true;
+          break;
+        }
+      }
+
+      if (!AnyStoreInStartBB) {
+        // No store in start BB. Add the lifetime marker as defining block to
+        // avoid undef.
+        Info.DefiningBlocks.insert(Info.DefiningBlocks.begin(), LifetimeStartBB);
```

Comment:
1. This was not intentional, good catch! Fixed by including a "single store" condition in the guarding if statement.
2. I'll take a closer look at this tomorrow, and will also try to come up with a dedicated test.

Reply:
Alright, I refreshed my memory and did a bit of stepping through use cases of this.
I updated the code with a larger comment that explains what's going on in detail. Let me just quote the relevant code:
```
    // Calculate the set of read and write-locations for each alloca.  This is
    // analogous to finding the 'uses' and 'definitions' of each variable.
    Info.AnalyzeAlloca(AI);

    // Determine whether this alloca only has one store *before* potentially
    // adding a lifetime.start block as another definition. This allows to
    // rewrite such an alloca efficiently if possible but still benefit from
    // correct lifetime information should the single store not dominate all
    // loads.
    const bool IsSingleStoreAlloca = Info.DefiningBlocks.size() == 1;

    // For allocas that have at least one store, make sure the block that holds
    // the lifetime.start call is treated as a "definition". This is important
    // in cases where lifetime is restricted to a loop and not all loads are
    // dominated by stores. When walking up the CFG during ComputeLiveInBlocks,
    // this new "definition" prevents the value from being considered live in
    // in the loop header, which in turn prevents the value to be live across
    // the entire loop through a phi with undef as input from the preheader.
    // Ignore allocas that are only used in one block, even if the lifetime
    // start marker is in a different one, since those cases are trivial.
    if (LifetimeStartBB && !Info.DefiningBlocks.empty() && !Info.OnlyUsedInOneBlock) {
      bool AnyStoreInStartBB = false;
      for (BasicBlock *BB : Info.DefiningBlocks) {
        if (BB == LifetimeStartBB) {
          AnyStoreInStartBB = true;
          break;
        }
      }

      if (!AnyStoreInStartBB) {
        // No store in start BB. Add the lifetime marker as defining block.
        Info.DefiningBlocks.insert(Info.DefiningBlocks.begin(), LifetimeStartBB);
      }
    }
```

To answer your second point, the lifetime start block itself is added to `LiveInBlocks`, but its predecessors are not. This prevents ComputeLiveInBlocks from visiting blocks in which the lifetime of the object hasn't started. No load is required in any of the visited blocks for the algorithm to create phi nodes. And it's not checked again whether a defining block actually contains a store.
I think this is only relevant inside loops where previously the loop preheader would be visited even if the lifetime start block was inside the loop, causing a phi node in the header with undef coming from the preheader. This can later not be undone since the information where lifetime started has effectively been altered to a location outside the loop.

I added the following unit test which exercises exactly the two situations you mention. s.x is only stored in the conditional and loaded in the subsequent block, which isn't dominated by the store (thus `rewriteSingleStoreAlloca` fails, so we still need the LifetimeStartBB even if we hit that path - I've now allowed both things to coexist by checking `Info.DefiningBlocks.size() == 1` before adding the new def block). Without the fix, ComputeLiveInBlocks will add all blocks up to the entry to `LiveInBlocks`, and subsequently the IDF identifies the loop header as a location where `QueuePhiNode` has to insert a phi.
```
struct MyStruct {
  int x;
};

RWStructuredBuffer<MyStruct> g_rwbuf : register(u0);

void expensiveComputation(int i, int x)
{
  // We'd prefer this to not be inlined to better illustrate things but that's
  // not possible without more effort.
  g_rwbuf[i].x = x;
}

export
int conditional_struct_init_in_loop(int n, int c1)
{
  int res = n;

  for(int i=0; i<n; ++i) {
    expensiveComputation(i, i*c1); // s mut not be considered live here.

    MyStruct s;

    // Initialize struct conditionally.
    // NOTE: If some optimization decides to flatten the if statement or if the
    //       computation could be hoisted out of the loop, the phi with undef
    //       below will be replaced by the non-undef value (which is a valid
    //       "specialization" of undef).
    if (c1 < 0)
      s.x = g_rwbuf[i - c1].x;

    res = s.x; // i or undef.
  }

  return res; // Result is n if loop wasn't executed, n-1 if it was.
}
```

FWIW, I think it's really interesting that this is missing from LLVM even at top-of-tree, but maybe the issue doesn't appear anywhere as pronounced as in the DXR use case.

Sorry for the wall-of-text :).